File: src/eQuantic.UI.Components/Accordion.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>One <see cref="Accordion"/> section: a title row and the content it expands to.</summary>
public sealed record AccordionItem(string Title)
{
    public VisualNode? Content { get; init; }
}

/// <summary>
/// The design system's Accordion (wave 3b): stacked sections whose 48dp header rows toggle their
/// content — the chevron flips, open sections show on the line-height grid, hairline dividers
/// between sections (the List convention). Open state is INTERNAL (like Menu): single-open by
/// default, <see cref="Multiple"/> keeps independent sections.
/// v1 fences: height animation (expand rides the state-transition system later), controlled mode,
/// disabled sections.
/// </summary>
public sealed class Accordion : StatefulComponent
{
    private int _openSingle = -1;
    private readonly HashSet<int> _openMulti = new();

    public Accordion(IReadOnlyList<AccordionItem> items, int openIndex = -1)
    {
        Items = items;
        _openSingle = openIndex;
    }

    public IReadOnlyList<AccordionItem> Items { get; private set; }
    public bool Multiple { get; init; }

    public override void AdoptConfig(UiComponent next)
    {
        if (next is Accordion fresh) Items = fresh.Items;
    }

    private bool IsOpen(int index) => Multiple ? _openMulti.Contains(index) : _openSingle == index;

    private void Toggle(int index)
    {
        SetState(() =>
        {
            if (Multiple)
            {
                if (!_openMulti.Add(index)) _openMulti.Remove(index);
            }
            else
            {
                _openSingle = _openSingle == index ? -1 : index;
            }
        });
    }

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var column = new Column(gap: 0) { Width = SizeValue.Fill };

        for (var i = 0; i < Items.Count; i++)
        {
            var item = Items[i];
            var index = i;
            var open = IsOpen(i);

            var header = new Row(gap: Space.S2) { Cross = CrossAlign.Center, Width = SizeValue.Fill };
            header.Add(new Text(item.Title, TypeRole.Label));
            header.Add(new Flexible(new Spacer()));
            header.Add(new Icon(open ? Icons.ChevronUp : Icons.ChevronDown, IconSize.Sm, theme.TextSecondary));

            column.Add(new Pressable(new Box(new BoxStyle
            {
                Height = 48,
                Width = SizeValue.Fill,
                Padding = EdgeInsets.Symmetric(Space.S3, 0),
                Hover = new StyleDiff { Background = theme.SurfaceSubtle },
            }, header), () => Toggle(index)));

            if (open && item.Content is { } content)
            {
                column.Add(new Box(new BoxStyle
                {
                    Width = SizeValue.Fill,
                    Padding = new EdgeInsets(Space.S3, 0, Space.S3, Space.S3),
                }, content));
            }

            if (i < Items.Count - 1) column.Add(new Divider());
        }

        return column;
    }
}

File: src/eQuantic.UI.Runtime/src/shared/components/Accordion.ts
import { Box, BoxStyle, BuildContext, Column, Divider, EdgeInsets, Flexible, Icon, Pressable, Row, SharedStatefulComponent, SizeValue, Spacer, StyleDiff, Text, UiComponent } from "../runtime-exports";

export class Accordion extends SharedStatefulComponent {
    _openSingle: number = -1;
    _openMulti: any = new Set();
    declare items: any;
    declare multiple: boolean;
    constructor(items?: any, openIndex: any = -1, props?: any) {
        super(props);
        if (items !== undefined) this.items = items;
        this.items = items;this._openSingle = openIndex;
    }

    build(context: BuildContext) {
        let theme = context.theme;let column = new Column(0, { width: SizeValue.fill });for (let i = 0; i < this.items.length; i++) {let item = this.items[i];let index = i;let open = this.isOpen(i);let header = new Row(8, { cross: 'center', width: SizeValue.fill, height: SizeValue.fill });header.add(new Text(item.title, 'label'));header.add(new Flexible(new Spacer()));header.add(new Icon(open ? 'chevronUp' : 'chevronDown', 16, theme.textSecondary));column.add(new Pressable(new Box(new BoxStyle({ height: 48, width: SizeValue.fill, padding: EdgeInsets.symmetric(12, 0), hover: new StyleDiff({ background: theme.surfaceSubtle }) }), header), () => this.toggle(index)));let content; if (open && ((item.content != null) && (content = item.content, true))) {column.add(new Box(new BoxStyle({ width: SizeValue.fill, padding: new EdgeInsets(12, 0, 12, 12) }), content));}if (i < this.items.length - 1) column.add(new Divider());}return column;
    }

    adoptConfig(next: UiComponent) {
        let fresh; if ((next instanceof Accordion && (fresh = next, true))) this.items = fresh.items;
    }

    isOpen(index: number) {
        return this.multiple ? this._openMulti.has(index) : this._openSingle === index;
    }

    toggle(index: number) {
        this.setState(() => {if (this.multiple) {if (!this._openMulti.add(index)) this._openMulti.delete(index);} else {this._openSingle = this._openSingle === index ? -1 : index;}});
    }

}

